plugins/inputs/firehose/firehose_request.go

```diff
@@ -0,0 +1,156 @@
+	}
+	return nil
+}
+
+func (r *firehoseRequest) validate() error {
+	requestID := r.req.Header.Get("x-amz-firehose-request-id")
+	if requestID == "" {
+		r.responseStatusCode = http.StatusBadRequest
+		return errors.New("x-amz-firehose-request-id header is not set")
+	}
+
+	// Check if content length is not over 64 MB.
+	if r.req.ContentLength > int64(64*1024*1024) {
+		r.responseStatusCode = http.StatusRequestEntityTooLarge
+		return fmt.Errorf("content length too large in request %s", requestID)
+	}
+
+	// Check if the requested HTTP method is allowed.
+	isAcceptedMethod := false
+	for _, method := range allowedMethods {
+		if r.req.Method == method {
+			isAcceptedMethod = true
+			break
+		}
+	}
+	if !isAcceptedMethod {
+		r.responseStatusCode = http.StatusMethodNotAllowed
+		return fmt.Errorf("forbidden method, %s, in request %s", r.req.Method, requestID)
+	}
+
+	contentType := r.req.Header.Get("content-type")
+	if contentType != "application/json" {
+		r.responseStatusCode = http.StatusBadRequest
+		return fmt.Errorf("unaccepted content type, %s, in request %s", contentType, requestID)
+	}
+
+	contentEncoding := r.req.Header.Get("content-encoding")
+	if contentEncoding != "" && contentEncoding != "gzip" {
+		r.responseStatusCode = http.StatusBadRequest
+		return fmt.Errorf("unaccepted content encoding, %s, in request %s", contentEncoding, requestID)
+	}
+
+	err := r.extractBody()
+	if err != nil {
+		return err
+	}
+
+	if requestID != r.body.RequestID {
+		r.responseStatusCode = http.StatusBadRequest
+		return errors.New("requestId in the body does not match the value of the request header, x-amz-firehose-request-id")
+	}
+
+	return nil
+}
+
+func (r *firehoseRequest) extractBody() error {
+	encoding := r.req.Header.Get("content-encoding")
+	switch encoding {
+	case "gzip":
+		g, err := gzip.NewReader(r.req.Body)
+		if err != nil {
+			r.responseStatusCode = http.StatusBadRequest
+			return fmt.Errorf("unable to decode body - %s", err.Error())
+		}
+		defer g.Close()
+		err = json.NewDecoder(g).Decode(&r.body)
+		if err != nil {
+			r.responseStatusCode = http.StatusBadRequest
+			return err
+		}
+	default:
+		defer r.req.Body.Close()
+		err := json.NewDecoder(r.req.Body).Decode(&r.body)
+		if err != nil {
+			r.responseStatusCode = http.StatusBadRequest
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *firehoseRequest) decodeData() ([][]byte, bool) {
+	// decode base64-encoded data and return them as a slice of byte slices
+	decodedData := make([][]byte, 0)
+	for _, record := range r.body.Records {
+		data, err := base64.StdEncoding.DecodeString(record.EncodedData)
+		if err != nil {
+			return nil, false
+		}
+		decodedData = append(decodedData, data)
+	}
+	return decodedData, true
+}
+
+func (r *firehoseRequest) sendResponse(res http.ResponseWriter) error {
+	responseBody := struct {
+		RequestID    string `json:"requestId"`
+		Timestamp    int64  `json:"timestamp"`
+		ErrorMessage string `json:"errorMessage,omitempty"`
+	}{
```

Comment:
Please declare this structure as you do with the request etc.

Reply:
done